bsms.mediawiki

```diff
@@ -0,0 +1,169 @@
+
+====Signer====
+
+The Signer is a participating member in the multisig. Its responsibilities include providing its key record -- which contains an Extended Public Key (XPUB) -- to the Coordinator, verifying that its XPUB is included in the descriptor record and persisting the descriptor record in its storage.
+
+===Setup Process===
+
+====Round 1====
+
+=====Coordinator=====
+
+* The Coordinator creates a multisig wallet creation session. The Coordinator constructs the multisig script and its policy parameters, such as the total number of signers and the required number of signatures (<tt>M</tt> and <tt>N</tt>).
+* The session should expire after some time period determined by the Coordinator, e.g., 24 hours. The timeout allows the encryption key to have lower entropy.
+* If encryption is enabled, the Coordinator distributes a secret <tt>TOKEN</tt> to each Signer over a secure channel. The Signer can use the <tt>TOKEN</tt> to derive an <tt>ENCRYPTION_KEY</tt>. Refer to the Encryption section below for details on the <tt>TOKEN</tt>, the key derivation function and the encryption scheme. Depending on the use case, the Coordinator can decide whether to share one common <tt>TOKEN</tt> for all Signers, or to have one per Signer.
+* If encryption is disabled, <tt>TOKEN</tt> is set to <tt>0</tt>, and all the encryption/decryption steps below can be skipped.
+
+=====Signer=====
+
+* The Signer initiates a new secure multisig setup session by setting the <tt>TOKEN</tt>. The Signer derives an <tt>ENCRYPTION_KEY</tt> from the <tt>TOKEN</tt>. The Signer can keep the session open until a different value for the <tt>TOKEN</tt> is set.
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Optionally, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line is a text description of the key, 80 characters maximum. The fourth line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first three lines.  The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted.
```

Comment:
It makes sense for the SIG to cover the TOKEN, to link the SIG to the session. But should it be included in the encrypted data to be sent ?

Coordinator can decrypt only with the key derived from correct TOKEN, so it can prepend it to the record before checking the signature.

Reply:
Again, not a cryptographer.  But something I noticed here is that the key record is signed, and then the signed message is encrypted, where usually the protocol is to have messages be encrypted, and then the encrypted messages are signed.  Does this create any issue?

Reply:
If the xpub is encrypted, you won't be able to check the signature before the decryption, so you can't do Encrypt-then-MAC with SIG. you could do MAC-then-Encrypt (the sig inside the cyphertext) or MAC-and-Encrypt (you need to decrypt before checking MAC).

But SIG is only used in one direction, I think it is better to leave it as a separate level, just an authentication of the fact that the signer has the privkey for the xpub, and that it willingly participates in the session represented by the TOKEN.

For authenticating the encryption of the whole record, and HMAC can be used, as in https://github.com/nunchuk-io/bips/pull/1#discussion_r583418860, and it can be used in both directions, uniformly.

Reply:
Makes sense, thanks